gnovm/cmd/gno/depgraph.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/gnolang/gno/gnovm/pkg/gnoenv"
+	"github.com/gnolang/gno/gnovm/pkg/gnomod"
+	"github.com/gnolang/gno/tm2/pkg/commands"
+	osm "github.com/gnolang/gno/tm2/pkg/os"
+)
+
+type depGraphCfg struct {
+	verbose        bool
+	rootDir        string
+	output         string
+	multipleGraphs bool
+}
+
+func newDepGraphCmd(io commands.IO) *commands.Command {
+	cfg := &depGraphCfg{}
+
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "depgraph",
+			ShortUsage: "depgraph [flags] <package> [<package>...]",
+			ShortHelp:  "generates dependency graphs for the specified packages",
+		},
+		cfg,
+		func(_ context.Context, args []string) error {
+			return execDepGraph(cfg, args, io)
+		},
+	)
+}
+
+func (c *depGraphCfg) RegisterFlags(fs *flag.FlagSet) {
+	rootDir := gnoenv.RootDir()
+	fs.BoolVar(&c.verbose, "v", false, "verbose output when lintning")
+	fs.StringVar(&c.rootDir, "root-dir", rootDir, "clone location of github.com/gnolang/gno (gno tries to guess it)")
+	fs.StringVar(&c.output, "o", "depgraph", "output (file if single graph, dir if multiple graphs)")
+	fs.BoolVar(&c.multipleGraphs, "m", false, "make a separate graph for each package")
+}
+
+func execDepGraph(cfg *depGraphCfg, args []string, io commands.IO) error {
+	if len(args) < 1 {
+		return flag.ErrHelp
+	}
+
+	var (
+		verbose        = cfg.verbose
+		rootDir        = cfg.rootDir
+		output         = cfg.output
+		multipleGraphs = cfg.multipleGraphs
+		allPkgs        gnomod.PkgList
+	)
+
+	// first we walk through all the folders and collect all packages
+	for _, arg := range args {
+		pkgs, err := gnomod.ListPkgs(arg)
+		if err != nil {
+			return fmt.Errorf("error in parsing gno.mod: %w", err)
+		}
+
+		allPkgs = append(allPkgs, pkgs...)
+	}
+
+	// make one big graph (eg. for the entire examples/ dir)
+	if !multipleGraphs {
+		nodeData := ""  //nodes
+		graphData := "" //edges
+
+		// we need subgraphs to make columns in the layout and for colors
+
+		// subgraph for .../p/...
+		nodeData += "subgraph {\nrank=same\n"
+		for _, pkg := range allPkgs {
+			if strings.Contains(pkg.Name, "gno.land/p") {
+				nodeData += "\"" + pkg.Name + "\" [color=\"blue\"]\n"
+			}
+		}
+
+		// subgraph for .../r/...
+		nodeData += "}\nsubgraph {\nrank=same\n"
+		for _, pkg := range allPkgs {
+			if strings.Contains(pkg.Name, "gno.land/r") {
+				nodeData += "\"" + pkg.Name + "\" [color=\"red\"]\n"
+			}
+		}
+		nodeData += "}"
+
+		for _, pkg := range allPkgs {
+			err := buildGraphData(pkg, allPkgs, make(map[string]bool), make(map[string]bool), &graphData)
+
+			if err != nil {
+				return fmt.Errorf("error in building graph: %w", err)
+			}
+		}
```

Comment:
The whole purpose of passing the visited map to this method is to keep track of already visited packages. You need to move the declaration of the visited map above the for loop to ensure there will be no duplicates while iterating between packages.

```suggestion
		visited := make(map[string]bool)
		for _, pkg := range allPkgs {
			err := buildGraphData(pkg, allPkgs, visited, make(map[string]bool), &graphData)

			if err != nil {
				return fmt.Errorf("error in building graph: %w", err)
			}
		}
```

Reply:
Doh! Yeah, I was thinking about solving that with a different structure, but this is simpler.